cert-ldap-login-module-common/src/main/java/org/esbtools/auth/ldap/LdapRolesProvider.java

```diff
@@ -59,29 +61,51 @@
  *
  */
 public class LdapRolesProvider implements RolesProvider {
-    private final Logger LOGGER = LoggerFactory.getLogger(LdapRolesProvider.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(LdapRolesProvider.class);
 
-    private String searchBase;
+    private final String searchBase;
 
-    private LdapConfiguration ldapConfiguration;
+    private final LdapConfiguration ldapConfiguration;
+
+    private final LDAPConnection ldapConnection;
 
     // Connection pool needs to be a singleton
+    /**
+     * @{code null} until {@link #connectIfNeeded()} is called and successful.
+     */
     private LDAPConnectionPool connectionPool;
+    private volatile LDAPException connectionException;
```

Comment:
It is odd to have an exception stored volatile.

Reply:
Some other thread might have assigned an exception while another thread goes to read it, we want to avoid reading an earlier failure or worse seeing null. It's not a huge deal if it's not immediately seen but because the variable won't be checked that often I figured better to be correct.